RavenDB-16880 Support PowerBI replace and filtering

omerlagziel opened the pull request.

Commit:
RavenDB-16880 Create a constructor for PgColumn

Commit:
RavenDB-16880 Replace for RQL queries without projection works, but not when there's more than 1 replace on a single column

Commit:
RavenDB-16880 Major refactor to PBIFetchQuery code to support nested queries

Commit:
RavenDB-16880 Multiple nested SQL finally works

Commit:
RavenDB-16880 Implemented ordered projection of selected columns

Commit:
RavenDB-16880 Organize code

Commit:
RavenDB-16880 Add proper support to generic fetch queries

Commit:
RavenDB-16880 Make replace work for multiple columns including id(), fix limit bug

Commit:
RavenDB-16880 Refactor code to avoid duplicate code

Commit:
RavenDB-16880 Support replace for RQLs with simple projection

Commit:
RavenDB-16880 Implement replace for all RQL types (except for edge cases documented in code)

Commit:
RavenDB-16880 Implement filters with the limitation of not being able to filter replaced columns (should filter first then replace in Power Query Editor)

Commit:
RavenDB-16880 Fix todos

Commit:
RavenDB-16880 Fix order logic to be more robust

Commit:
RavenDB-16880 Created classes for postgres types

Commit:
RavenDB-16880 Move PgTypeConverter logic into PgType classes

Commit:
RavenDB-16880 Moved TypeModifier logic into the PgTypes

arekpalinski commented on src/PgRvn.Server/HardcodedQuery.cs:
```diff
@@ -0,0 +1,144 @@
+
+namespace PgRvn.Server
+{
+    public class HardcodedQuery : PgQuery
```
What is actually the meaning of "HardcodedQuery"?

ayende replied:
There are several queries that we just recognize. 
For example, `npgsql` type loading, or Power BI inits. 
We don't bother parsing the SQL, just do the minimum to recognize them.

arekpalinski commented on src/PgRvn.Server/HardcodedQuery.cs:
```diff
@@ -0,0 +1,144 @@
+                throw new InvalidOperationException("Didn't expect more than one SQL statement in queryString, got: " + sqlStatements.Count);
+            }
+
+            foreach (var stmt in sqlStatements)
```
I don't see any usage of `stmt`. Is that correct?

omerlagziel replied:
Yeah I also just noticed it, apparently TSQLParser doesn't return anything meaningful in `stmt.ToString()` and I couldn't find anything else to give me the statement.

So for now I just got rid of the `foreach` and used the full `QueryString`. Maybe we can find a better parser later.

arekpalinski commented on src/PgRvn.Server/HardcodedQuery.cs:
```diff
@@ -0,0 +1,144 @@
+            foreach (var stmt in sqlStatements)
+            {
+                var powerBiMatch = PowerBIConfig.TypesQuery;
+                if (QueryString.Equals(powerBiMatch))
```
Wouldn't be more clear to compare it directly: `if (QueryString.Equals(PowerBIConfig.TypesQuery))`? Do we need `powerBiMatch` variable? What is the purpose of having it?

omerlagziel replied:
Done

arekpalinski commented on src/PgRvn.Server/HardcodedQuery.cs:
```diff
@@ -0,0 +1,144 @@
+                // TODO: Treat the results like any other data source and acknowledge the results format
+                var resultsFormat = GetDefaultResultsFormat();
+
+                powerBiMatch = "select character_set_name from INFORMATION_SCHEMA.character_sets";
```
Can we move it to a const like other other queries?

omerlagziel replied:
Done

arekpalinski commented on src/PgRvn.Server/MessageBuilder.cs:
```diff
@@ -353,6 +353,16 @@ private void WriteInt16(short value, Span<byte> buffer, ref int pos)
         private void WriteByte(byte value, Span<byte> buffer, ref int pos)
         {
+            // TODO: Work on this for every function
+            if(pos +1 > buffer.Length)
```
Minor - formatting
```suggestion
            if (pos + 1 > buffer.Length)
```

omerlagziel replied:
Done

arekpalinski commented on src/PgRvn.Server/MessageBuilder.cs:
```diff
@@ -353,6 +353,16 @@ private void WriteInt16(short value, Span<byte> buffer, ref int pos)
+                var newBuf = _bufferOwner.Memory.Span;
+                buffer.CopyTo(newBuf);
+                buffer = newBuf;
+                oldOwner.Dispose();
```
Better to wrap it inside `using` instead of explicitly calling `Dispose()`:

```
using (var oldOwner = _bufferOwner)
{ 
   ...
}
```

This way if something throw an exception you're guaranteed that the `Dispose()` will be called

omerlagziel replied:
Done

arekpalinski commented on src/PgRvn.Server/PowerBI/PBIAllCollectionsQuery.cs:
```diff
@@ -19,9 +19,10 @@ public PBIAllCollectionsQuery(string queryText, int[] parametersDataTypes, IDocu
-            if (queryText.Replace("\r", string.Empty) == tableQuery)
+            const string tableQuery = "select TABLE_SCHEMA, TABLE_NAME, TABLE_TYPE\nfrom INFORMATION_SCHEMA.tables\nwhere TABLE_SCHEMA not in ('information_schema', 'pg_catalog')\norder by TABLE_SCHEMA, TABLE_NAME";
+
+            queryText = queryText.Replace("\r\n", "\n").Replace("\r", "\n");
```
I think you use `Replace("\r\n", "\n").Replace("\r", "\n")` in multiple places already. Maybe create an extension method for that e.g. `NormalizeLineEndings()`

omerlagziel replied:
Done

arekpalinski commented on src/PgRvn.Server/PowerBI/PBIFetchQuery.cs:
```diff
@@ -2,210 +2,363 @@
                 {
-                    var collection = rqlMatch.Groups["collection"];
+                    var jsSelectKeys = rql.Groups["js_keys"];
+                    var jsSelectValues = rql.Groups["js_vals"];
```
It's a bit hard to me to follow what's going on here. What is exactly `js_keys` and `js_vals`? My intuition says me that it might be:

```
... select 
{
   FirstName: u.FirstName,
   LastName: u.LastName
}
```

then `js_keys` is `FirstName, LastName` and `js_vals` is `u.FirstName, u.LastName`. Is that correct?

omerlagziel replied:
That's correct, hopefully the new version of the code is more clear

arekpalinski commented on src/PgRvn.Server/PowerBI/PBIFetchQuery.cs:
```diff
@@ -2,210 +2,363 @@
+                for (var i = matches.Count - 1; i >= 0; i--)
+                {
+                    var match = matches[i];
+                    alreadyPopulated = alreadyPopulated || i != matches.Count - 1;
```
`i != matches.Count - 1` what does it mean? maybe create a local variable for that with meaningful name

omerlagziel replied:
Done - Simplified code

arekpalinski commented on src/PgRvn.Server/PowerBI/PBIFetchQuery.cs:
```diff
@@ -2,210 +2,363 @@
-            return false;
+            var limit = matches[0].Groups["limit"];
+
+            Console.WriteLine(">> RQL:\n" + newRql + "\n");
```
Please add `// TODO remove me` or something similar so you won't forget to delete it eventually

omerlagziel replied:
Removed it for now, I'll keep it in mind if I ever add it back

arekpalinski commented on src/PgRvn.Server/PowerBI/PBIFetchQuery.cs:
```diff
@@ -2,210 +2,363 @@
-                if (replaceDestColumnsSet.TryGetValue(columnName, out var value))
+        private static void PopulateProjectionFields(
+            Match match, 
+            IDictionary<string, string> projectionFields, 
```
Maybe let's add `ref` to  `projectionFields` parameter to indicate explicitly that it's modified inside this method

omerlagziel replied:
Added

arekpalinski commented on src/PgRvn.Server/PowerBI/PBIFetchQuery.cs:
```diff
@@ -2,210 +2,363 @@
-                projection += ", ";
+                if (replace[i].Value.Length != 0)
+                {
+                    destColumnVal = $"({destColumnVal}).toString()" +
```
How does it work if the column isn't a string? E.e. we have numeric value, we replace it but we no longer return numeric but a string, yes?

ayende replied:
`replace` operates on strings

arekpalinski replied:
okay

omerlagziel replied:
Yes, I made it so if the user attempts to replace a non-string value the field will be converted to a string. If that's the wanted behavior please resolve this thread.

arekpalinski replied:
That's good behavior IMO

arekpalinski commented on src/PgRvn.Server/PowerBI/PBIFetchQuery.cs:
```diff
@@ -2,210 +2,363 @@
+            }
+        }
+
+        private static string GenerateColumnValue(string column, string alias)
```
We call it to generate it for the projection using JS syntax, right? Then maybe include that info in the method name - e.g. `GenerateRqlJsProjectionColumnValue()` (or something better that you could think about)

The idea is that another dev reading that code would know immediately what is it about.

omerlagziel replied:
Done

arekpalinski commented on src/PgRvn.Server/PowerBI/PBIFetchQuery.cs:
```diff
@@ -2,210 +2,363 @@
+            var selectIndex = select.Success ? select.Index : (whereIndex + whereLength);
+
+            // Insert alias clause if doesn't exist
+            if (!alias.Success)
```
Minor - as a convention we always use `if (alias.Success == false)` instead of `if (!alias.Success)`. It's easier to read it

omerlagziel replied:
Done

arekpalinski commented on src/PgRvn.Server/PowerBI/PBIFetchQuery.cs:
```diff
@@ -2,210 +2,363 @@
+
+            // Insert new where clause
+            // Note: Filtering of projected columns is not supported
+            var fullNewWhere = "";
```
Can we use StringBuilder here (and in other places) and use its `Append()` method instead of manual string concatenation? StringBuilder offers better performance. You might find it a useful reading; https://www.infoworld.com/article/3616600/when-to-use-string-vs-stringbuilder-in-net-core.html

ayende replied:
Also important, put `\t` and line breaks as needed, so the RQL is readable.

omerlagziel replied:
Done

arekpalinski commented on src/PgRvn.Server/RqlQuery.cs:
```diff
@@ -91,128 +92,85 @@ private ICollection<PgColumn> GenerateSchema()
                 _hasIncludes = true;
-                Columns["is_include()"] = new PgColumn
+
+                if (Columns.TryGetValue("is_include()", out var includesColumn))
```
What is `is_include()`? Where does it come from?

ayende replied:
can be removed

omerlagziel replied:
Done

Commit:
RavenDB-16880 Major refactor of message handling logic to make state management more easy to follow, fix accuracy of protocol implementation and add more robust error handling

Commit:
RavenDB-16880 Add buffer resizing functionality to MessageBuilder

Commit:
RavenDB-16880 Limit parameter size to 1MB

Commit:
RavenDB-16880 Fix CR comments

Commit:
RavenDB-16880 Document members

ayende commented on src/PgRvn.Server/HardcodedQuery.cs:
```diff
@@ -0,0 +1,145 @@
+
+        public void Parse(bool allowMultipleStatements)
+        {
+            var sqlStatements = TSQLStatementReader.ParseStatements(QueryString);
```
So we are using the SQL Parser? I thought we skipped it and used regexes?

omerlagziel replied:
Only to recognize the number of statements in the SQL. Unfortunately `TSQLParser` doesn't even let us get the query string of each statement separately. 

If you want I can get rid of it and stop checking if there's more than 1 statement when not allowed (e.g. Extended Protocol only allows 1 query per transaction).

ayende replied:
I would rather avoid having a dependency on the SQL parser, yes.
And supporting just one is fine.

omerlagziel replied:
Done

ayende commented on src/PgRvn.Server/HardcodedQuery.cs:
```diff
@@ -0,0 +1,145 @@
+                powerBiMatch = "select character_set_name from INFORMATION_SCHEMA.character_sets";
+                if (QueryString.Equals(powerBiMatch))
+                {
+                    _result = new PgTable
```
here as well

omerlagziel replied:
done

ayende commented on src/PgRvn.Server/Messages/MessageReader.cs:
```diff
@@ -360,9 +369,23 @@ private async Task<byte> ReadByteAsync(PipeReader reader, CancellationToken toke
+
+        private void SkipBytes(ReadOnlySequence<byte> readBuffer, PipeReader reader, int length)
+        {
+            var sequence = readBuffer.Slice(0, length);
```
Don't think this will work. If the size is 32 MB, for example, you'll error here.

omerlagziel replied:
I tested this with Npgsql and a `bytea` parameter of 52 MB and it seems to work

ayende commented on src/PgRvn.Server/Messages/MessageReader.cs:
```diff
@@ -360,9 +369,23 @@ private async Task<byte> ReadByteAsync(PipeReader reader, CancellationToken toke
         {
             var sequence = readBuffer.Slice(0, length);
+
+            var buffer2 = ArrayPool<byte>.Shared.Rent(length);
```
huh? Why do you have this and still retain the old buffer?
Also, where do you intend to return it?

Also, note that if you rent an array, it may come up *bigger* than expected.

omerlagziel replied:
Not sure why that was added, removing it

ayende commented on src/PgRvn.Server/PgConfig.cs:
```diff
@@ -23,122 +25,23 @@ public static class PgConfig
-                    DataTypeSize = sizeof(int),
-                    FormatCode = PgFormat.Text //PgFormat.Binary
-                }
+                new PgColumn("nspname", 0, PgName.Default, PgFormat.Text),
```
*far* more readable.

ayende commented on src/PgRvn.Server/RqlQuery.cs:
```diff
@@ -91,128 +93,85 @@ private ICollection<PgColumn> GenerateSchema()
 
-            if (_result.Includes.Count > 0 ||
-                sample.TryGet("@metadata", out BlittableJsonReaderObject metadata) && metadata.TryGet("@id", out string _))
+            if (sample.TryGet("@metadata", out BlittableJsonReaderObject metadata) && metadata.TryGet("@id", out string _) ||
```
We can drop the support for includes, I think. That is no longer relevant for our use cases.

omerlagziel replied:
Done

ayende commented on src/PgRvn.Server/Types/PgBit.cs:
```diff
@@ -0,0 +1,44 @@
+        public override short Size => sizeof(byte);
+        public override int TypeModifier => 1;
+
+        public override byte[] ToBytes(object value, PgFormat formatCode)
```
We probably should avoid `byte[]`, it will force us to allocate, need to think about another interface.

omerlagziel replied:
Changed to ReadOnlyMemory\<byte\>

Commit:
RavenDB-16880 Continue fixing both PRs

Commit:
RavenDB-16880 Add CSV to PgTable handling

Commit:
RavenDB-16880 Fix CsvToPg and implement support for Npgsql versions 3.2.3 and above.

Commit:
RavenDB-16880 Add full documented support for all versions of Npgsql above 3.2.3

Commit:
RavenDB-16880 Add better formatting of generated RQL

Commit:
RavenDB-16880 Use StringBuilder

Commit:
RavenDB-16880 Major refactor - fix bug where connection ends after non fatal error

The reason for this refactor is that we needed to be able to get an instance of the message even if continuing reading it will cause a PgErrorException.
We needed the instance to later do message.HandleError which will result in different behaviour for ExtendedProtocolMessage types (set transaction state to failed) and regular Message types.

arekpalinski commented on src/PgRvn.Server/Exceptions/PgErrorException.cs:
```diff
@@ -1,6 +1,6 @@
-namespace PgRvn.Server
+namespace PgRvn.Server.Exceptions
 {
     public class PgErrorException : Exception
```
Let's make sure that all our custom exceptions are serializable by adding `[Serializable]` attribute and proper construtors. See here: https://stackoverflow.com/questions/94488/what-is-the-correct-way-to-make-a-custom-net-exception-serializable

Example of such exception in RavenDB code: https://github.com/ravendb/ravendb/blob/v5.2/src/Raven.Server/Exceptions/MissingAttachmentException.cs

ayende replied:
FWIW - pretty sure that this no longer meaningful with the .NET core shift

ayende replied:
Also, note that this is used internally, so we would never actually serialize this.

arekpalinski replied:
OK. No need to do anything then

Commit:
RavenDB-16880 Use ArrayPool to handle small parameters and regular allocation for big ones

Commit:
RavenDB-16880 Fix lifespan of MessageHandler to transaction lifespan

Commit:
RavenDB-16880 Removed TSQL Dependency

Commit:
RavenDB-16880 Remove patch RQLs support
